### PR TITLE
Endpoint for uploading a file and learn

### PR DESCRIPTION
**Pull request by subins2000**
Added `/learn/upload` endpoint for files to be uploaded and learnt

### CODE REVIEW

**Comment by subins2000** [comment]
The last commit makes `/learn` also under auth. Currently, `api.varnamproject.com` allows anyone to submit a word for learning. `/learn` is public endpoint. But, people can misuse it, similar with the file upload. So, making it both under an auth endpoint.

File: http_handlers.go [hunk 1]
```diff
@@ -280,6 +280,94 @@ func handleLearn(c echo.Context) error {
...
+	form, err := c.MultipartForm()
+	if err != nil {
+		return err
+	}
+
+	langCode := c.FormValue("lang")
+	files := form.File["files"]
+
+	_, ok := trainChannel[langCode]
+	if !ok {
```
Comment (joicemjoseph) [inline http_handlers.go]: can be
```go
if_, ok := trainChannel[langCode]; !ok {
```

File: http_handlers.go [hunk 2]
```diff
@@ -280,6 +280,94 @@ func handleLearn(c echo.Context) error {
...
+		app.log.Printf("unknown language requested to learn: %s", langCode)
+		return echo.NewHTTPError(http.StatusBadRequest, "unable to find language to train")
+	}
+
+	if len(files) == 0 {
+		return echo.NewHTTPError(http.StatusBadRequest, "No file uploaded")
```
Comment (joicemjoseph) [inline http_handlers.go]: add log here

Comment (joicemjoseph) [reply http_handlers.go]: log errors everywhere in handlers.

File: http_handlers.go [hunk 3]
```diff
@@ -280,6 +280,94 @@ func handleLearn(c echo.Context) error {
...
+			return echo.NewHTTPError(http.StatusInternalServerError, err.Error())
+		}
+		defer src.Close()
+
+		learnFilePath := getSyncMetadataDir() + "/learn-" + langCode + "-" + file.Filename
+
+		// Destination
+		dst, err := os.Create(learnFilePath)
+		if err != nil {
+			return echo.NewHTTPError(http.StatusInternalServerError, err.Error())
```
Comment (joicemjoseph) [inline http_handlers.go:335]: add log

File: http_handlers.go [hunk 4]
```diff
@@ -280,6 +280,94 @@ func handleLearn(c echo.Context) error {
...
+	if len(files) == 0 {
+		return echo.NewHTTPError(http.StatusBadRequest, "No file uploaded")
+	}
+
+	// Copy files first
+	for _, file := range files {
+		// Source
+		src, err := file.Open()
+		if err != nil {
+			return echo.NewHTTPError(http.StatusInternalServerError, err.Error())
```
Comment (joicemjoseph) [inline http_handlers.go:328]: log here

File: http_handlers.go [hunk 5]
```diff
@@ -280,6 +280,94 @@ func handleLearn(c echo.Context) error {
...
+	}
+
+	// Learn
+	for _, file := range files {
+		learnFilePath := getSyncMetadataDir() + "/learn-" + langCode + "-" + file.Filename
+
+		learnFromFile(langCode, learnFilePath)
+	}
+
+	return c.HTML(http.StatusOK, "success")
```
Comment (joicemjoseph) [inline http_handlers.go]: why html, json is fine.

File: http_handlers.go [hunk 6]
```diff
@@ -280,6 +280,94 @@ func handleLearn(c echo.Context) error {
...
+		}
+		defer dst.Close()
+
+		// Copy
+		if _, err = io.Copy(dst, src); err != nil {
+			return echo.NewHTTPError(http.StatusInternalServerError, err.Error())
+		}
+	}
+
+	learnFromFile := func(langCode, fileToLearn string) {
```
Comment (joicemjoseph) [inline http_handlers.go]: it seems like a bad practice. Refactor into a separate function.

**Review by joicemjoseph — CHANGES_REQUESTED** [review]

File: http_handlers.go (hunk 6 above)
Comment (subins2000) [reply http_handlers.go]: @joicemjoseph Thought of it as a callbacks like in JavaScript, it's not preferred in Go eh ?

File: server.go [hunk 7]
```diff
@@ -34,9 +34,11 @@ func initHandlers(app *App, enableInternalApis bool) *echo.Echo {
 	e.GET("/rtl/:langCode/:word", handleReverseTransliteration)
 	e.GET("/meta/:langCode:", handleMetadata)
 	e.GET("/download/:langCode/:downloadStart", handleDownload)
-	e.POST("/learn", handleLearn)
 	e.GET("/languages", handleLanguages)
 	e.GET("/status", handleStatus)
+
+	e.POST("/learn", authUser(handleLearn))
```
Comment (joicemjoseph) [inline server.go]: should we put learn endpoint behind authentication ?

**Review by joicemjoseph — APPROVED** [review]

**Pull request merged**